plugin-server/src/cdp/hog-watcher.ts

```diff
@@ -0,0 +1,321 @@
+const REDIS_KEY_DISABLED_HISTORY = `${BASE_REDIS_KEY}/disabled_history`
+const REDIS_TIMEOUT_SECONDS = 5
+
+// NOTE: We ideally would have this in a file but the current build step doesn't handle anything other than .ts files
+const LUA_TOKEN_BUCKET = `
+local key = KEYS[1]
+local now = ARGV[1]
+local cost = ARGV[2]
+local poolMax = ARGV[3]
+local fillRate = ARGV[4]
+local expiry = ARGV[5]
+local before = redis.call('hget', key, 'ts')
+
+-- If we don't have a timestamp then we set it to now and fill up the bucket
+if before == false then
+  local ret = poolMax - cost
+  redis.call('hset', key, 'ts', now)
+  redis.call('hset', key, 'pool', ret)
+  redis.call('expire', key, expiry)
+  return ret
+end
+
+-- We update the timestamp if it has changed
+local timeDiffSeconds = now - before
+
+if timeDiffSeconds > 0 then
+  redis.call('hset', key, 'ts', now)
+else
+  timeDiffSeconds = 0
+end
+
+-- Calculate how much should be refilled in the bucket and add it
+local owedTokens = timeDiffSeconds * fillRate
+local currentTokens = redis.call('hget', key, 'pool')
+
+if currentTokens == false then
+  currentTokens = poolMax
+end
+
+currentTokens = math.min(currentTokens + owedTokens, poolMax)
+
+-- Remove the cost and return the new number of tokens
+if currentTokens - cost >= 0 then
+  currentTokens = currentTokens - cost
+else
+  currentTokens = -1
+end
+
+redis.call('hset', key, 'pool', currentTokens)
+redis.call('expire', key, expiry)
+
+-- Finally return the value - if it's negative then we've hit the limit
+return currentTokens
+`
+
+export enum HogWatcherState {
+    healthy = 1,
+    degraded = 2,
+    disabledForPeriod = 3,
+    disabledIndefinitely = 4,
+}
+
+export type HogWatcherFunctionState = {
+    state: HogWatcherState
+    tokens: number
+    rating: number
+}
+
+type WithCheckRateLimit<T> = {
+    checkRateLimit: (key: string, now: number, cost: number, poolMax: number, fillRate: number, expiry: number) => T
+}
+
+type HogWatcherRedisClientPipeline = Pipeline & WithCheckRateLimit<number>
+
+type HogWatcherRedisClient = Omit<Redis, 'pipeline'> &
+    WithCheckRateLimit<Promise<number>> & {
+        pipeline: () => HogWatcherRedisClientPipeline
+    }
+
+export class HogWatcher {
+    constructor(private hub: Hub) {}
+
+    private rateLimitArgs(id: HogFunctionType['id'], cost: number) {
+        const nowSeconds = Math.round(now() / 1000)
+        return [
+            `${REDIS_KEY_TOKENS}/${id}`,
+            nowSeconds,
+            cost,
+            this.hub.CDP_WATCHER_BUCKET_SIZE,
+            this.hub.CDP_WATCHER_REFILL_RATE,
+            this.hub.CDP_WATCHER_TTL,
+        ] as const
+    }
+
+    private async runRedis<T>(fn: (client: HogWatcherRedisClient) => Promise<T>): Promise<T | null> {
+        // We want all of this to fail open in the issue of redis being unavailable - we'd rather have the function continue
+        const client = await this.hub.redisPool.acquire()
+
+        client.defineCommand('checkRateLimit', {
```

Comment:
Could we do this in the constructor?

Reply:
Annoyingly no. It has to be done per-client and the client is created in the line above (and then cached). 

I could modify it so that the client creation code includes it but that would now be done for the whole of plugin-server which feels bad. 

I checked the source and it doesn't issue anything, it just defines something in memory (with a small hash check of the lua code) so didn't _seem_ to be a noticeable perf impact.

Reply:
can the HogWatcher cache the client on construction?

Reply:
I really don't think its worth it at this point (I really dont want to affect anything else in the plugin server doing redis). Its a simple assignment so there's really no cost here.

Reply:
I don't see how constructing a client and setting up the command in the constructor of the `HogWatcher` is different to constructing one and setting up the command on every call of this function, with respect to impacting the rest of the plugin server?